lib/wibox/init.lua

```diff
@@ -61,6 +63,72 @@ function wibox:find_widgets(x, y)
     return self._drawable:find_widgets(x, y)
 end
 
+-- This function is called on every redraw after the background has been drawn,
+-- but before the actual widget was drawn. We use this to draw an anti-aliased
+-- clipping shaped (the edge between the border and the background is
+-- anti-aliased)
+local function draw_hook(self, cr)
+    local shape = self._shape
+    if not shape then return end
+
+    local geo = self:geometry()
+    cr:set_source(gcolor(self.border_color))
+
+    -- We want to invert the shape (draw to anything that is not inside of it).
+    -- To fake this, we use EVEN_ODD and a rectangle containing the full
+    -- surface. Pixels in the border will be inside the big rectangle, so they
+    -- are contained once in the path. Pixels in the "content area" are
+    -- additionally contained in the shape, so twice.
+    cr:set_fill_rule(cairo.FillRule.EVEN_ODD)
+    cr:rectangle(0, 0, geo.width, geo.height)
+    shape(cr, geo.width, geo.height)
+
+    cr:fill()
+end
+
+function wibox:_apply_shape()
+    local shape = self._shape
+
+    if not shape then
+        self.shape_bounding = nil
+        self.shape_clip = nil
+        return
+    end
+
+    local geo = self:geometry()
+    local bw = self.border_width
+
+    -- First handle the bounding shape (things including the border)
+    local img = cairo.ImageSurface(cairo.Format.A1, geo.width + 2*bw, geo.height + 2*bw)
+    local cr = cairo.Context(img)
+
+    shape(cr, geo.width + 2*bw, geo.height + 2*bw)
```

Comment:
> Should we really try to make X11 do something that it is not meant to do?

Yes, because everybody else does it and because for rounded rectangles, hexagons and octogons (the 3 most popular use cases), it has no issue.

Also, here's something to help you test:

```lua
local cairo = require( "lgi" ).cairo

-- Shape with hardcoded values or values based on only 1 axis don't necessarely
-- support linear radial scaling from the shape centroid. This forces the shape
-- to always be called with the same with and height or the result will miss
-- part of its border on one side
local function hardcoded_shape(cr, w, h)
    --actually, I didn't test it, but I think it triggers the bug
    cr:move_to(0,10)
    cr:line_to(10, 0)
    cr:line_to(w, 0)
    cr:line_to(w, h-10)
    cr:line_to(w-10, h)
    cr:line_to(0, h)
    cr:line_to(0, 10)
    cr:close_path()
end

-- Set a checkboard wallpaper
local img = cairo.ImageSurface(cairo.Format.ARGB32, 20, 20)
local cr = cairo.Context(img)
cr:set_source_rgb(1,1,1)
cr:paint()
cr:set_source_rgb(0.5,0.5,0.5)
cr:rectangle(0,0,10,10)
cr:rectangle(10,10,10,10)
cr:fill()
local pat = cairo.Pattern.create_for_surface(img)
gears.wallpaper.tiled(img, 1)

local w5 = wibox {
    visible = true,
    width = 200,
    height = 200,
    x= 50,
    y=50,
    ontop = true,
    shape = function(cr,w,h) gears.shape.octogon(cr,w,h, 30) end,
    bg = "#ff0000",
    border_width = 4,
    border_color = "#00ff00",
}
local w6 = wibox {
    visible = true,
    width = 200,
    height = 200,
    x= 270,
    y=50,
    ontop = true,
    shape = hardcoded_shape,
    bg = "#ff0000",
    border_width = 4,
    border_color = "#00ff00",
}
local w7 = wibox {
    visible = true,
    width = 200,
    height = 200,
    x= 490,
    y=50,
    ontop = true,
    shape = gears.shape.circle,
    bg = "#ff0000",
    border_width = 4,
    border_color = "#00ff00",
}
local w8 = wibox {
    visible = true,
    width = 200,
    height = 200,
    x= 50,
    y=270,
    ontop = true,
    shape = function(cr,w,h) gears.shape.octogon(cr,w,h, 30) end,
    bg = "#ff0000",
    border_width = 0,
}
local w9 = wibox {
    visible = true,
    width = 200,
    height = 200,
    x= 270,
    y=270,
    ontop = true,
    shape = hardcoded_shape,
    bg = "#ff0000",
    border_width = 0,
}
local w10 = wibox {
    visible = true,
    width = 200,
    height = 200,
    x= 490,
    y=270,
    ontop = true,
    shape = gears.shape.circle,
    bg = "#ff0000",
    border_width = 0,
}
```